packages/flutter/lib/src/widgets/animated_switcher.dart

```diff
@@ -230,153 +272,142 @@ class AnimatedSwitcher extends StatefulWidget {
+      controller: controller,
+      animation: animation,
+      builder: widget.transitionBuilder,
+    );
+    if (animate) {
+      controller.forward();
+    } else {
+      assert(_outgoingEntries.isEmpty);
+      controller.value = 1.0;
+    }
   }
 
-  _AnimatedSwitcherChildEntry _newEntry({
+  _ChildEntry _newEntry({
+    @required Widget child,
+    @required AnimatedSwitcherTransitionBuilder builder,
     @required AnimationController controller,
     @required Animation<double> animation,
   }) {
-    final _AnimatedSwitcherChildEntry entry = _AnimatedSwitcherChildEntry(
-      widgetChild: widget.child,
-      transition: KeyedSubtree.wrap(
-        widget.transitionBuilder(
-          widget.child,
-          animation,
-        ),
-        serialNumber++,
-      ),
+    final _ChildEntry entry = _ChildEntry(
+      widgetChild: child,
+      transition: KeyedSubtree.wrap(builder(child, animation), _childNumber),
       animation: animation,
       controller: controller,
     );
     animation.addStatusListener((AnimationStatus status) {
       if (status == AnimationStatus.dismissed) {
         setState(() {
-          _removeExpiredChild(entry);
+          assert(mounted);
+          assert(_outgoingEntries.contains(entry));
+          _outgoingEntries.remove(entry);
+          _markChildWidgetCacheAsDirty();
         });
         controller.dispose();
       }
     });
     return entry;
   }
 
-  void _removeExpiredChild(_AnimatedSwitcherChildEntry child) {
-    assert(_previousChildren.contains(child));
-    _previousChildren.remove(child);
-    _markChildWidgetCacheAsDirty();
-  }
-
-  void _retireCurrentChild() {
-    assert(!_previousChildren.contains(_currentChild));
-    _currentChild.controller.reverse();
-    _previousChildren.add(_currentChild);
-    _markChildWidgetCacheAsDirty();
-  }
-
   void _markChildWidgetCacheAsDirty() {
-    _previousChildWidgetCache = null;
+    _outgoingWidgets = null;
   }
 
-  void _addEntry({@required bool animate}) {
-    if (widget.child == null) {
-      if (animate && _currentChild != null) {
-        _retireCurrentChild();
-      }
-      _currentChild = null;
-      return;
-    }
-    final AnimationController controller = AnimationController(
-      duration: widget.duration,
-      vsync: this,
+  void _updateTransitionForEntry(_ChildEntry entry) {
+    entry.transition = KeyedSubtree(
+      key: entry.transition.key,
+      child: widget.transitionBuilder(entry.widgetChild, entry.animation),
     );
-    if (animate) {
-      if (_currentChild != null) {
-        _retireCurrentChild();
-      }
-      controller.forward();
-    } else {
-      assert(_currentChild == null);
-      assert(_previousChildren.isEmpty);
-      controller.value = 1.0;
-    }
-    final Animation<double> animation = CurvedAnimation(
-      parent: controller,
-      curve: widget.switchInCurve,
-      reverseCurve: widget.switchOutCurve,
+  }
+
+  void _rebuildChildWidgetCacheIfNeeded() {
```

Comment:
You might rename this to `_rebuildOutgoingWidgetsIfNeeded` since the widget cache is no longer the name of what it operates on.

Reply:
done